disasm-test/Main.hs

```diff
@@ -318,119 +361,200 @@ cube = TS.mkCUBE
+diffCmp file1 file2 = do
+  let assertF = liftIO . assertFailure . unlines
+  (code, stdout, stderr) <- liftIO $
+    Proc.readCreateProcessWithExitCode (Proc.proc "diff" ["-u", file1, file2]) ""
+  case code of
+    ExitFailure _ -> assertF ["diff failed", stdout, stderr]
+    ExitSuccess   ->
+      if stdout /= "" || stderr /= ""
+      then assertF ["non-empty diff", stdout, stderr]
+      else do Details det <- gets showDetails
+              when det $ liftIO
+                $ mapM_ putStrLn ["success: empty diff: ", file1, file2]
+
+
+-- Assembles the specified .ll file to bitcode, then disassembles it with
+-- llvm-dis.  Also parses the bitcode with this library (effectively llvm-disasm)
+-- and prints the difference between the parsed version and the .ll file.
+-- Returns the library parsed version and the serialized AST from the library.
 
 processLL :: FilePath -> FilePath -> TestM (FilePath, Maybe FilePath)
 processLL pfx f = do
-  Details dets <- gets showDetails
-  when dets $ liftIO $ putStrLn (showString f ": ")
-  X.handle logError $
-    withFile (assembleToBitCode pfx f)  $ \ bc   ->
-    withFile (disasmBitCode pfx bc) $ \ norm -> do
-      (parsed, ast) <- processBitCode pfx bc
-      when dets $ liftIO $ do
-        ignore (Proc.callProcess "diff" ["-u", norm, parsed])
-        putStrLn ("successfully parsed " ++ show f)
-      return (parsed, ast)
+  Details det <- gets showDetails
+  when det $ liftIO $ putStrLn (showString f ": ")
+  X.handle logError
+    $ withFile (assembleToBitCode pfx f)
+    $ parseBC pfx
   where
     logError (ParseError msg) =
       liftIO $ assertFailure $ unlines
       $ "failure" : map ("; " ++) (lines (formatError msg))
 
+parseBC :: FilePath -> FilePath -> TestM (FilePath, Maybe FilePath)
+parseBC pfx bc = do
+  withFile (disasmBitCode pfx bc) $ \ norm -> do
+    (parsed, ast) <- processBitCode pfx bc
+    Details dets <- gets showDetails
+    when dets $ liftIO $ do
+      -- Informationally display if there are differences between the llvm-dis
+      -- and llvm-disasm outputs, but no error if they differ.
+      ignore (Proc.callProcess "diff" ["-u", norm, parsed])
+      putStrLn ("successfully parsed " ++ show pfx ++ " bitcode")
+    return (parsed, ast)
+
+----------------------------------------------------------------------
+-- Compiler->Assembly->Disassembly tests
+
+-- The compilation tests ensure that the clang version-specific generated .bc
+-- file can be reasonably parsed by this library.  This is a parallel to the
+-- assemblyCube-driven tests, but starts with a C source file.  One distinction
+-- is that the .ll used for the assemblyCube is typically representative of a
+-- specific LLVM version, and while it is assembled and disassembled by newer
+-- versions of LLVM tools, it will never introduce any newer element, whereas the
+-- clang-generated bitcode will contain version-current output which might have
+-- newer elements and ordering.
```

Comment:
How fragile is this? I worry that new Clang versions might introduce minor differences in LLVM instructions that, while not changing the user-observable behavior of the program, would nevertheless cause this `diff` test to fail. Do you expect that we will have to add a lot of additional, version-specific `.ll` expected files over time?

Reply:
There are definitely issues discovered using this method that are not seen with the `.ll` sources.  This may require adding new version-specific .ll target expected files in the future. The primary goal here is to be able to parse any bitcode file generated by clang (or any other llvm front-end); we are currently measuring  the success of that by determining if we can emit the parsed form in a format that the `llvm-as` tool will successfully parse back into bitcode, and then repeating that cycle to ensure that the two forms match.  I agree we may have to deal with format change issues, but at present I don't see a better way to test this, and adding these clang tests has revealed at least 5 bugs (not all are present in this PR: I wanted to establish the testing framework and then try to file them individually after this is in place; the two noted here are already present from existing `.c` files in the test corpus).

Reply:
OK. My feeling on the matter is that we should keep the size of these `*.c` programs as small as possible to reduce the chance of future Clang versions compiling things in an ever-so-slightly different way and breaking the diff. It might be worth making a note of this general guideline in the [`llvm-disasm` `README`](https://github.com/GaloisInc/llvm-pretty-bc-parser/blob/df5b7d63cd6d7b06ecd13d9ce8eb189add2ab90a/disasm-test/README.md).

Reply:
I'm in agreement, and I updated the descriptive text in the disasm-text (there didn't seem to be a great spot in the README).